keras_cv/layers/regularization/squeeze_excite.py

```diff
@@ -0,0 +1,73 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+from tensorflow.keras import layers
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class SqueezeAndExciteBlock2D(layers.Layer):
+    """
+    Implements Squeeze and Excite block as in
+    [Squeeze-and-Excitation Networks](https://arxiv.org/pdf/1709.01507.pdf).
+
+    Args:
+            filters: Number of input and output filters. The number of input and
+                    output filters is same.
+            ratio: Ratio for bottleneck filters. Number of bottleneck filters =
+                    filters * ratio. Defaults to 0.25.
+    Usage:
+
+    ```python
+    # (...)
+    input = tf.ones((1, 5, 5, 16), dtype=tf.float32)
+    x = tf.keras.layers.Conv2D(16, (3, 3))(input)
+    output = keras_cv.layers.SqueezeAndExciteBlock(16)(x)
+    # (...)
+    ```
+    """
+
+    def __init__(self, filters, ratio=0.25, **kwargs):
+        super().__init__(**kwargs)
+
+        self.filters = filters
+
+        if ratio <= 0.0 or ratio >= 1.0:
+            raise ValueError(f"`ratio` should be a float between 0 and 1. Got {ratio}")
+
+        if filters <= 0 or not isinstance(filters, int):
+            raise ValueError(f"`filters` should be a positive integer. Got {filters}")
+
+        self.ratio = ratio
+        self.bottleneck_filters = int(self.filters * self.ratio)
+
+        self.global_average_pool = layers.GlobalAveragePooling2D(keepdims=True)
+        self.squeeze_conv = layers.Conv2D(
+            self.bottleneck_filters,
+            (1, 1),
+            activation="relu",
+        )
+        self.excite_conv = layers.Conv2D(self.filters, (1, 1), activation="sigmoid")
```

Comment:
Do we need to support `squeeze_activation`, `excite_activation` parameters?

Reply:
![image](https://user-images.githubusercontent.com/64411306/176258786-23bbabb7-11a3-4bb2-9cc3-f5e53a8eaee7.png)
This implementation is based on the original paper (https://arxiv.org/pdf/1709.01507.pdf), so I think ReLU and Sigmoid are more or less mandatory. But would like to hear your and @sayakpaul's thoughts on the same.

Reply:
MobileNetv3 uses SE blocks with HardSigmoid - we should support this level of parameterization whether or not the original paper used that or not.  

Keep in mind ML research moves fast, and while research is great to reference it should not be treated as a hard source of truth.  Another examples of this is that ML research has a tendency to give things strange names that stick for a year or two, then are left behind in favor of a more idiomatic name.  This has happened numerous times, and as such we pick names based on idiomaticness rather than their name in the research - excluding some niche cases like RandAugment.

Reply:
Makes sense. Thanks for the detailed explanation. I'll make the changes and re-request a review.

Reply:
`squeeze_activation` is a better naming than `excite_activation` IMO.

Reply:
@sayakpaul I agree but I guess we need to have both of those (one for squeeze conv and other for excite conv).

Reply:
> However, in that case, in the get_config method squeeze_activation or excite_activation will be keras Layers, which won't be serializable. 

Why won't they be serializable? 

I think we could support strings that can be directly passed to `activation` or we could allow users to pass functions rather than layers.

Reply:
Definitely. For one moment, I got confused with `tf.keras.initializers.*`. We can pass a class under it (such as `VarianceScaling`) to arguments like `kernel_initializer`. But this is not a layer like you mentioned.

Reply:
> > I think we could support strings that can be directly passed to activation or we could allow users to pass functions rather than layers.
> 
> Sure, since the functional versions of activations are a strict superset (I believe) of the layer versions of the same, it would be favorable.
> 
> Just as a side note, it is interesting to see if one can get layers to be serializable whose arguments are layer objects. I tried it just now but it isn't as simple as it seems 😅

Can you paste a snippet of the issue here?  I think layers to the constructor should be serializable - it should call their get_config() under the hood if I remember correctly.

Reply:
Sure. 

Code which raises error:
```py
    def get_config(self):
        config = {
            "filters": self.filters,
            "ratio": self.ratio,
            "squeeze_activation": self.squeeze_activation, # instance of, say, tf.keras.activations.ReLU()
            "excite_activation": self.excite_activation,
        }
        base_config = super().get_config()
        return dict(list(base_config.items()) + list(config.items()))
```
error log: shared on https://codeshare.io/3AMwy6


My workaround:
```py
def get_config(self):
        squeeze_activation_serialized = (
            serialize(self.squeeze_activation)
            if hasattr(self.squeeze_activation, "__call__")
            else self.squeeze_activation
        )
        excite_activation_serialized = (
            serialize(self.excite_activation)
            if hasattr(self.squeeze_activation, "__call__")
            else self.excite_activation
        )

        config = {
            "filters": self.filters,
            "ratio": self.ratio,
            "squeeze_activation": squeeze_activation_serialized,
            "excite_activation": excite_activation_serialized,
        }
        base_config = super().get_config()
        return dict(list(base_config.items()) + list(config.items()))

```